### ISSUE TITLE
C6.4e: source-kernel-preserving representation v4 + exhaustive finite-domain equivalence protocol

### ISSUE THREAD

**Issue by Alajmah** [ElephantRock/Continuity-Aware-Distributed-Inference#126]
Fifth bounded C6.4 representation-revision slice under #102, following C6.4d PR #125 / merge `4b937cd65c7731981a04524279d43a74f3032dc8`.

## Why this slice exists

C6.4d rejected representation v3 before fresh adequacy because its canonical scalar polynomial evaluator did not preserve the pinned Vidur/sklearn binary64 evaluation kernel within the frozen 8-ULP transcription budget. The corrected full-domain fence found only numerical-transcription discrepancies (max absolute error on the order of `1e-15` ms), but they are real contract failures and v3 was correctly rejected.

C6.4d also established an important methodological fact: the prior source execution has already materialized the complete finite Vidur prediction domain used by the TP1/PP1 profile (except identities that were already separately invalidated by superseded eager materialization). Therefore there is no scientifically honest way to manufacture a new *untouched* scalar prefill/transfer boundary inside that same finite lookup domain. Re-labeling another subset as fresh would be false.

The replacement protocol therefore changes the validation form, not the source or oracle: representation v4 must preserve the source numerical kernel and C6.4f must prove equivalence **exhaustively over the complete finite declared request domain**. Because the domain is finite and fully enumerated, this is stronger than sampling another holdout and makes no generalization claim beyond the declared source domain.

## Representation v4 to freeze here

Representation ID:

```text
cadi.c6.cost-representation.v4.source-ordered-polynomial+source-linear-algebra+decode-step-affine
```

For every Vidur `PolynomialFeatures + LinearRegression` component, v4 preserves rather than canonicalizes away the source evaluation semantics:

```text
ordered input feature names
exact upstream PolynomialFeatures `powers_` row order
constant/bias rows exactly where produced by upstream
exact regression coefficient vector in that same order
exact regression intercept
exact selected upstream hyperparameters
explicit evaluation-kernel identifier
accepted C6.4b protocol/source fingerprints
```

The scientific evaluation kernel is the already-frozen C6.4b numerical substrate and must reproduce the source linear algebra without sklearn estimator objects:

```text
CPython 3.12.14
numpy 1.26.4
OpenBLAS 0.3.23.dev
OPENBLAS_CORETYPE=Haswell
threading_layer=pthreads
num_threads=1
```

The v4 evaluator must construct the transformed float64 row in the exact frozen source power order and perform the regression operation with the same float64 matrix/vector arithmetic contract as the source `LinearRegression.predict`, including intercept addition. It may depend on NumPy for this scientific projection; it must not depend on sklearn objects at inference/evaluation time. No coefficient folding, monomial reordering, `math.fsum`, clipping, smoothing, or post-hoc numerical repair is allowed.

Cold-prefill request composition, recompute semantics, transfer token/byte domain, carried C6.3 decode primitive/composition, state bytes, and 80-GiB memory-capacity mechanics remain unchanged unless a source-code identity proves the prior transcription itself was wrong. No C6.4b/C6.4d residual may be used to choose coefficients or arithmetic order.

## Exhaustive equivalence protocol to freeze here

C6.4f will evaluate v4 against the exact pinned Vidur source under the accepted fixed numerical substrate on **every public request axis in the declared finite profile domain**, not a sampled boundary:

```text
cold prefill input tokens: 1..4096
point-to-point transfer predictor tokens: 1..4096
transfer public bytes: predictor_tokens * 8192
recompute: same prefill function, therefore covered by prefill exhaustively
single-token decode: carried C6.3 held-out-adequate primitive, unchanged
multi-token decode: unchanged analytic per-step composition
state/memory: unchanged source-derived algebra
```

For A100 and H100 independently, C6.4f must compare source request-level seconds with v4 request-level seconds on all 4096 prefill and all 4096 transfer axes. The frozen 8-ULP transcription budget remains applicable to source-equivalence values; the existing 5% MAPE / 10% max-APE adequacy thresholds are retained as reportable aggregate guards but are not used to excuse any source-equivalence violation.

A profile passes only if every finite-domain source-equivalence point passes the transcription budget, all outputs are finite/non-negative, source/protocol/profile fingerprints are valid, carried decode remains adequate, and state/memory provenance is intact. Parent pass requires both A100 and H100.

## Boundary accounting

The C6.4c second boundary remains permanently:

```text
INVALIDATED_BY_SUPERSEDED_PREDICTOR_MATERIALIZATION
```

It must never again be called fresh or untouched. C6.4e does **not** create a replacement holdout boundary because the complete declared source lookup domain has already been consumed by C6.4d source-equivalence work. C6.4f may include all historical axes in an exhaustive domain proof, but only as exhaustive equivalence evidence; it must not describe them as validation holdouts or new measurements.

## Anti-leakage / no-repair rules

C6.4e/C6.4f fails review if any of the following occurs:

- a new subset of the already-consumed finite source domain is described as fresh or untouched;
- v4 arithmetic order is selected by minimizing C6.4d residuals rather than by preserving the pinned upstream execution semantics;
- coefficients/intercepts/hyperparameters are changed after exhaustive results are observed;
- source predictions are clipped, rounded, smoothed, refit, or silently extrapolated;
- the numerical substrate/source fingerprint drifts silently;
- modeled P-SRC2 values are described as direct A100/H100 physical measurements;
- exhaustive equivalence is used to claim validity outside the frozen source domain.

## Exit for C6.4e

Close this slice only after:

- representation-v4 schema and evaluator contract are committed;
- source-order/bias/intercept/kernel invariants are fail-closed and regression-tested;
- the exhaustive C6.4f domain/oracle is frozen in code/spec **before** any v4 source-equivalence run is executed;
- no source timing/prediction is evaluated merely to choose or modify v4 mechanics;
- bounded technical/scientific review passes;
- exact-head deterministic-core CI is green;
- SHA-fenced squash merge completes.

C6.4f then performs the exhaustive source proof. C6.4g simulator integration remains blocked unless C6.4f passes. C6.5 remains the final reproducibility/exit review.

**Comment by Alajmah** [comment]
Pre-timing source-code identity refinement (no Vidur model prediction was executed): representation v4 will also preserve the pinned request-level arithmetic order, not merely the component regression kernel. Pinned `vidur/entities/execution_time.py` computes cold-prefill `model_time` through ordinary binary64 additions in this exact nesting: attention-layer sum (`pre_proj + post_proj + rope + kv_cache_save + decode + attn_prefill + TP_comm + attn_norm`), MLP-layer sum (`up_proj + down_proj + act + TP_comm + mlp_norm`), then `attention + mlp + add`, multiply by layers, add PP communication, then `* 1e-3`. C6.4c/v3 used `math.fsum` over a flattened component list, which is semantically equivalent but not a source-exact numerical transcription. v4 will follow the pinned source nesting/order exactly. This refinement is justified directly by source blob `vidur/entities/execution_time.py@a5100f86b7e4d885ff62a5a66f845682a998add8`, before any v4 source-equivalence run.
